gpiozero/tones.py

```diff
@@ -0,0 +1,200 @@
+# vim: set fileencoding=utf-8:
+
+from __future__ import (
+    unicode_literals,
+    absolute_import,
+    print_function,
+    division,
+)
+str = type('')
+
+import re
+from collections import namedtuple
+try:
+    from math import log2
+except ImportError:
+    from .compat import log2
+
+
+class Tone(float):
+    """
+    Represents a frequency of sound in a variety of musical notations.
+
+    :class:`Tone` class can be used with the :class:`~gpiozero.TonalBuzzer`
+    class to easily represent musical tones. The class can be constructed in a
+    variety of ways. For example as a straight frequency in `Hz`_ (which is the
+    internal storage format), as an integer MIDI note, or as a string
+    representation of a musical note.
+
+    All the following constructors are equivalent ways to construct the typical
+    tuning note, `concert A`_ at 440Hz which is MIDI note #69):
+
+        >>> from gpiozero.tones import Tone
+        >>> Tone(440.0)
+        >>> Tone(69)
```

Comment:
Euw, I'm afraid I don't like this at all. Given that GpioZero is aimed at newcomers to Python, I think they'd find the fact that `Tone(60)` creates a midi-note of 60 and not a frequency of 60Hz very confusing.

Reply:
Hmm, it's a tricky one. My reasoning is that this is a bit like colorzero's Color constructor(s). The main constructor is essentially a convenience and part of being convenient is a trade-off regarding ambiguities. Color's ambiguity is admittedly tiny (does `Color(1, 0, 0)` mean full red, or a very very dim red?), and Tone's is *much* larger but then there's a much larger overlap in the domains it's trying to represent (frequencies and MIDI notes here).

The alternate constructors mitigate this to a degree by allowing the user to explicitly specify what they expect (i.e. `Tone.from_frequency(60)` will always be 60Hz, and `Tone.from_midi(60)` will always be middle C). Still, that doesn't answer the question of why I choose to prioritize MIDI for the 0..127 range.

This was mostly down to how much of the range could be represented. Let's say we prioritize MIDI (as I did). In this case 0..127 covers the entire MIDI range, while the frequency domain only gets 128Hz..20Khz, losing the bottom range. However, the alternative (prioritize the frequency domain) means MIDI gets ... erm ... note 0, and frequency gets 1Hz..20Khz. An alternative I did consider was permitting 0..20 for MIDI notes because as a rough rule of thumb, tones below 20Hz are inaudible. However, this still leaves the MIDI system almost unusable because the concert range of A0..G9 is represented by MIDI notes 21..127. So to leave the MIDI system even vaguely useful in the constructor means chopping out a (small) portion of the lower end of the frequency range.

I did, however, leave one get-out clause: floating-point numbers are *always* treated as frequencies. So `Tone(60)` is MIDI note 60, `Tone(60.0)` is 60Hz. Frankly though, I'm not sure if that's an even more egregious ambiguity (which is why, partly out of embarrassment, it's not prominently documented!).

That leads finally to the question: do we care about MIDI? Is it useful? I'd argue for new programmers it's *much* easier to deal with than the alternative systems. Each note is a simple integer, not a string that needs constructing or parsing. There's no worrying about the octave number changing between B and C, rather than the wrapping point of A and G (as with the otherwise human friendly notation system<sup>1</sup>), and there's no powers or logs to deal with in calculating the frequency of the next or prior semi-tone (as there is with frequencies). It's just simple addition and subtraction: perfect for programming without getting into the heavy maths.

So that in the end is why I went with prioritizing MIDI over the frequency range - I'd be most interested in any counter-arguments though (I frequently enjoy @lurch's reviews - they often lead me to consider my justifications for things or, as in this case, to formalize some of the choices I'd already run through in my head).

<sup>1</sup> Actually, now I think about it, the notation system *isn't* particularly human friendly; it's quite weird in some respects. It's actually *musician* friendly (a fraternity I still somewhat count myself among the ranks of, having passed grade 8 theory *many* years ago ... don't ask me to explain the plagal cadence though; I've forgotten most of it!)

Reply:
How about using kwargs like `Tone(frequency=440)` over `Tone.from_frequency(440)`?

Reply:
> How about using kwargs like Tone(frequency=440) over Tone.from_frequency(440)?

Ah, knew I'd forgotten something. Yes, I should add those back in (although it's barely less typing than just using the explicit constructor). Incidentally, what do you think of the (horrid) `Tone(60)` vs `Tone(60.0)`. Worth excising? I'd say yes, in light of keyword-args.

Reply:
...and/or we could borrow inspiration from #470 and allow `Tone('MIDI25')` and `Tone('60Hz')`? :wink: 

As already mentioned, I have **no** musical background, but having `Tone(127)` construct a midi-note and having `Tone(128)` construct a frequency, sounds like a big disconnect / ambiguity to me. If the current behaviour is to be kept, IMHO it ought to be much better documented in the docstring.

Reply:
...I guess another alternative might be to _only_ allow construction from kwargs, and treat `Tone(60)` as an error? :man_shrugging:

Reply:
Yes, I think that's what Dave suggested earlier ^ if I'm reading it correctly. If so, I think we're all in agreement.

Reply:
Actually I just suggested a warning in that case rather than a full-blown exception. My problem with an exception for ambiguous cases is that it'll blow up for almost *all* MIDI notes (1..127 as it's only 0 that can't be a frequency). A warning would have the same issue but would only report once (given warnings default filter), and wouldn't just terminate the user's script for wanting to use MIDI (or perhaps frequency!).

Still, lurch's concern is a reasonable one: what about users who simply want to play with frequencies (an interesting thing to do whether one is musical or not)?

Okay ... decision time:

1. I'll remove the really horrid int/float ambiguity handling. There's too much danger of someone doing arithmetic on an int, accidentally turning it into a float and suddenly getting a pure frequency instead of the MIDI note they expected and then having no clue why.

2. I'll introduce the warning for the ambiguous cases (numeric values between 1 and 127). The exception's too harsh (basically bans the use of MIDI notes entirely) and the warning, although it'll trigger for basically *all* MIDI notes, will only trigger on the first use due to the default suppression of repeat warnings. The warning message will include instructions for explicitly specifying the frequency and MIDI systems in case the user got the wrong one, and/or just wishes to suppress the warning.

3. I'll tweak the `tone` setter to permit any value that would validly construct a `Tone` instance. From the above, that'll also raise a warning in ambiguous cases.

4. I won't be adding in the extra `note`, `midi`, and `frequency` properties - partly because it makes for a rather bloated API, but also partly because I have some plans/vague musings for "note" (hint: what's the difference between a tone and a note?).

5. I'm still undecided on the removal of `play()`. So much for decision time ... Ben, you want to make a call on this?

Reply:
`play` should stay in play :)